doc/performance/0002-http-vs-rcp-client-performance-analysis.md

```diff
@@ -0,0 +1,94 @@
+```
+
+The goal of this document is to compare the performance of the TSP client before and after the change. Since we have an additional link in the communication chain, it is expected that the communication time will either be the same or increase. We would like to know how much is this increased, and how it affects the performance of the Theia Trace Extension.
+
+**Update**: It is confirmed that Theia no longer uses JSON-RCP for it front end (FE)-back end (BE) communication, but they use a faster technology for serializing JSON called `message pack` instead. So the actual implementation would be:
+
+```
+Theia FE <--Message Pack--> Theia BE <--TSP over HTTP--> Trace Server
+```
+
+## Test cases
+
+There were 3 test cases that were carried out, which simply involved:
+
+1. Measure the starting time
+2. Run the TSP function that we need to measure performance
+3. Measure the ending time
+4. Log the time difference to the console
+
+This were done manually simply because measuring the performance after the JSON-RPC patch requires a bit more effort, which involves `playwright testing`. This would be an interesting task to be done in the future if we decide that we need more detailed information regarding the performance post-JSON-RPC. For now, manual tests would be sufficient.
+
+The three functions that were tested are:
+
+1. tspClient.experimentOutputs(): Open 4 traces. I used the 4 traces in the 103 sample traces of traviz-labs. Then reload the extension to trigger the function. Make sure all trace tabs are closed. This has the least amount of data sent to the Theia BE with just 1 parameter.
+
+2. tspClient.fetchXYTree(): Open a histogram view. This requires a few more parameters sent to the Theia BE compare to the prior test case.
+
+3. tspClient.fetchTimeGraphStates(): Open a Flame Chart. I used the bad-trace sample that were used in the timeline-chart performance analysis. This requires the most amount of data to be sent to the Theia BE.
+
+## Results
+
+### Pre-JSON-RPC-patch
+
+|experimentOutputs() (ms)(*)| fetchXYTree() (ms)| fetchTimeGraphStates() (ms)
+|---|---|---|
+|472.5|90.7|100.8|
+|301|98.4|108.2|
+|465.3|95.7|139.2|
+|865.9|91.6|126.7|
+|905.1|103.6|201.8|
+
+On average we have:
+
+|experimentOutputs() (ms)(*)| fetchXYTree() (ms)| fetchTimeGraphStates() (ms)
+|---|---|---|
+|601.6|96.0|135.34|
+
+(*): For the case of fetchTimeGraphStates(), the number varies quite a bit. If we run more test attempts, sometimes we have outliers that are very small (24-35ms). However, it looks like most of the time, the range is 400-600ms.
+
+### Post-JSON-RPC-patch
+
+|experimentOutputs() (ms)(**)| fetchXYTree() (ms)| fetchTimeGraphStates() (ms)
+|---|---|---|
+|503.4|92.5 |292.8|
+|160.1|95.7|142.9|
+|116.5|97.9|128.5|
+|159.2|100.5|133.8|
+|175.9|91|123.8|
+
+On average we have:
+
+|experimentOutputs() (ms)(**)| fetchXYTree() (ms)| fetchTimeGraphStates() (ms)
+|---|---|---|
+|223.0|94.3|164.5|

```

Comment:
Averages have significance limits, as we know. Now, here, it is surprising that adding a json-rpc leg diminishes the latency overall. Including the aforementioned logging points might help narrow this issue down somewhat, in case their respective locations would be debatable.